### PR TITLE
Restricting MultiAppUserObjectTransfer to boundaries and blocks

### LINKED ISSUE #16241: Make MultiAppUserObjectTransfer boundary and block restrictable
## Reason

In more complicated setups, users want to transfer values from multiple user-objects into a single variable. These values are usually either boundary or subdomains restricted. Multiple transfers on a single variable currently overwrite the values previously set, because the transfer operates on the whole domain. This can be avoided by restricting the transfer.

## Design

Add `block` and `boundary` parameter into `MultiAppUserObjectTransfer` so that the transfer is localized just to the list of either boundaries or subdomains and avoids overwriting the values set previously by other transfers.

## Impact

Gain functionality. Better UX: easier input file setup (less aux variables needed to capture subdomain/boundary restricted values - not variables). Also easier visualization, since users can look at just one variable and see the whole field and don't have to switch between multiple ones.

### PR DESCRIPTION
**Pull request by andrsd**
Closes #16241

### CODE REVIEW

**Review by aeslaughter — COMMENTED** [review]
What are you not using the BlockRestrictable interface class?

**Comment by andrsd** [comment]
> What are you not using the BlockRestrictable interface class?

I started with that, but ran into the following problems: The block-/boundary-restrictable classes are designed to work with the mesh/variables of "this" `FEProblem`. So that works only for transfers from sub-app to master app. The other direction does not work, because (a) there can be multiple instances of the sub-apps, (b) at the time of creating the transfer the sub-app infrastructure is not in place yet. That is also the reason the block/boundary restriction is checked/inited/set every time we call `execute` (the same approach is done in other transfers that can be restricted to blocks/boundaries).

**Comment by aeslaughter** [comment]
Can you add that reason to the class comments please.

**Comment by andrsd** [comment]
Yeah, that's a good idea...

**Comment by andrsd** [comment]
The doxygen for this class is wrong - seems like a copy-paste from some other class, so I will fix that as well...

File: framework/include/transfers/MultiAppUserObjectTransfer.h [hunk 1]
```diff
@@ -43,4 +94,10 @@ class MultiAppUserObjectTransfer : public MultiAppConservativeTransfer
 
   /// whether to check the bounding box check or not
   const bool _skip_bbox_check;
+
+  /// Set of block ids this transfer is restricted to
+  std::set<SubdomainID> _blk_ids;
+
+  /// Set of the boundary ids
+  std::set<BoundaryID> _bnd_ids;
```
Comment (loganharbour) [inline framework/include/transfers/MultiAppUserObjectTransfer.h:102]: These really should be private.

File: framework/src/transfers/MultiAppUserObjectTransfer.C [hunk 2]
```diff
@@ -100,14 +109,29 @@ MultiAppUserObjectTransfer::execute()
...
-          MeshBase * mesh = NULL;
+          MooseMesh * mesh = NULL;
 
           if (_displaced_target_mesh && _multi_app->appProblemBase(i).getDisplacedProblem())
+            mesh = &_multi_app->appProblemBase(i).getDisplacedProblem()->mesh();
+          else
+            mesh = &_multi_app->appProblemBase(i).mesh();
+
+          _blk_ids.empty();
+          _bnd_ids.empty();
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppUserObjectTransfer.C:120]:
```suggestion
          _blk_ids.clear();
          _bnd_ids.clear();
```

File: framework/src/transfers/MultiAppUserObjectTransfer.C [hunk 3]
```diff
@@ -211,12 +247,28 @@ MultiAppUserObjectTransfer::execute()
...
 
       if (_displaced_target_mesh && to_problem.getDisplacedProblem())
-        to_mesh = &to_problem.getDisplacedProblem()->mesh().getMesh();
+        to_mesh = &to_problem.getDisplacedProblem()->mesh();
       else
-        to_mesh = &to_problem.mesh().getMesh();
+        to_mesh = &to_problem.mesh();
+
+      _blk_ids.empty();
+      _bnd_ids.empty();
```
Comment (loganharbour) [inline framework/src/transfers/MultiAppUserObjectTransfer.C:258]:
```suggestion
      _blk_ids.clear();
      _bnd_ids.clear();
```

**Review by loganharbour — CHANGES_REQUESTED** [review]
Minor changes.

**Comment by loganharbour** [comment]
The more that I think about it - why can the block/boundary IDs not be set in the constructor and then verified in `execute()`? Or maybe even earlier?

File: framework/src/transfers/MultiAppUserObjectTransfer.C (hunk 2 above)
Comment (andrsd) [reply framework/src/transfers/MultiAppUserObjectTransfer.C:120]: Good catch. Thanks!

File: framework/include/transfers/MultiAppUserObjectTransfer.h (hunk 1 above)
Comment (andrsd) [reply framework/include/transfers/MultiAppUserObjectTransfer.h:102]: Ok.

**Comment by andrsd** [comment]
It is also explained the class doxygen. Nobody cares about doco 🙃

**Review by loganharbour — APPROVED** [review]

**Pull request merged**